core/field.ts

```diff
@@ -1014,28 +1053,28 @@ export abstract class Field<T = any> implements IASTNodeLocationSvg,
    * @param validatedValue Validated value.
    * @returns New value, or an Error object.
    */
+  private processValidation_(newValue: T, validatedValue: Validation<T>): T
+      |Error;
   private processValidation_(
-      newValue: AnyDuringMigration,
-      validatedValue: AnyDuringMigration): AnyDuringMigration {
+      newValue: U|Un, validatedValue: ClassValidation<T, U>): T|Error|Un;
+  private processValidation_(newValue: T|U|Un, validatedValue: T|null|Un): T
+      |Error|Un {
     if (validatedValue === null) {
       this.doValueInvalid_(newValue);
       if (this.isDirty_) {
         this.forceRerender();
       }
       return Error();
     }
-    if (validatedValue !== undefined) {
-      newValue = validatedValue;
-    }
-    return newValue;
+    return validatedValue === undefined ? newValue as T | Un : validatedValue;
```

Comment:
> If `newValue` is undefined, that would be a class validation error.

The case where the return value could be `undefined` is if `doClassValidation_` is overwritten by a subclass. See my reply above where I talk about it more (https://github.com/google/blockly/pull/6639#discussion_r1048866050). I list a case where this could occur (and one where it could actually be valid). 

Thinking through that case, though, it's only valid when `T={somethingElse}|undefined`, so the `T` in `processValidation_`'s return type should take care of that. The invalid `doClassValidation_ ` override  case probably doesn't matter to know here since it's not being handled anyway when I assume it's defined at `const newClassValue = classValue!;`  in `setValue`.


> If that's the case, then instead of overloading the function, based on the guidance [here](https://www.typescriptlang.org/docs/handbook/2/functions.html#writing-good-overloads): "Always prefer parameters with union types instead of overloads when possible"

Thank you for linking the [TypeScript Guidance](https://www.typescriptlang.org/docs/handbook/2/functions.html#writing-good-overloads)! That's helpful to refer to here. The reason I overloaded the function here originally was that it allowed TS to know that `classValue` could be `undefined` but `localValue` could not. That'll no longer be the case, though.

> So we should go for what's clearest in intent over being exactly correct but verbose and needs a lot of studying to figure out IMO.

Sounds good! I certainly can get carried away with my type info 😆. It was more of an in-code documentation thing here when there were two distinct return cases for `processValidation_`.

The `ClassValidation` does its work when people override `doClassValidation_`. By this point, that distinction no longer matters.